src/ifcgeom/mapping/IfcGradientCurve.cpp

```diff
@@ -0,0 +1,76 @@
+/********************************************************************************
+ *                                                                              *
+ * This file is part of IfcOpenShell.                                           *
+ *                                                                              *
+ * IfcOpenShell is free software: you can redistribute it and/or modify         *
+ * it under the terms of the Lesser GNU General Public License as published by  *
+ * the Free Software Foundation, either version 3.0 of the License, or          *
+ * (at your option) any later version.                                          *
+ *                                                                              *
+ * IfcOpenShell is distributed in the hope that it will be useful,              *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 *
+ * Lesser GNU General Public License for more details.                          *
+ *                                                                              *
+ * You should have received a copy of the Lesser GNU General Public License     *
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.         *
+ *                                                                              *
+ ********************************************************************************/
+
+#include "mapping.h"
+#define mapping POSTFIX_SCHEMA(mapping)
+using namespace ifcopenshell::geometry;
+
+#ifdef SCHEMA_HAS_IfcGradientCurve
+
+taxonomy::ptr mapping::map_impl(const IfcSchema::IfcGradientCurve* inst) {
+	auto horizontal = taxonomy::cast<taxonomy::piecewise_function>(map(inst->BaseCurve()));
+	auto vertical = taxonomy::make<taxonomy::piecewise_function>();
+
+	auto segments = inst->Segments();
+
+	for (auto& segment : *segments) {
+		if (segment->as<IfcSchema::IfcCurveSegment>()) {
+			// @todo check that we don't get a mixture of implicit and explicit definitions
+			auto crv = map(segment->as<IfcSchema::IfcCurveSegment>());
+			if (crv->kind() == taxonomy::PIECEWISE_FUNCTION) {
+				auto seg = taxonomy::cast<taxonomy::piecewise_function>(crv);
+				vertical->spans.insert(vertical->spans.end(), seg->spans.begin(), seg->spans.end());
+			} else {
+				Logger::Error("Unsupported");
+				return nullptr;
+			}
+		} else {
+			Logger::Error("Unsupported");
+			return nullptr;
+		}
+	}
+
+	// @todo does this really make sense?
```

Comment:
We are making another assumption here - sort of - maybe it's just a distinction we need to be mindful of. 

For CT Alignment Geometry Segment, the representation is on the IfcCurveSegment whereas the other CTs put the representation on the IfcAlignment.

Reply:
> because it seems to require that the ParentCurve be 3D

Where do you see this constraint encoded? I think that would be an error. I briefly looked and IfcShapeRepresentation.RepresentationType=Segment seems to just select IfcCurveSegment without any constraints on the parent curve.

IFC is a community where clear choices are seldomly made. So I think the idea is you do everything. Individual representations for the segments is the most likely thing missing from these aspects. But you're right in the sense that due to the IFC schema flexibility, you never really know.

I've updated my diagram.

![instance-diagram dot](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/cf999255-8cdb-4770-8dfe-1f2feda67ffa)

<details>
<summary>dot code</summary>

```
digraph {

node [shape=rect]

IfcAlignment -> IfcRelNests_0[dir=back]
IfcRelNests_0 -> { IfcAlignmentHorizontal, IfcAlignmentVertical, IfcAlignmentCant }

IfcAlignmentHorizontal -> IfcRelNests_1[dir=back]
IfcRelNests_1 -> { IfcAlignmentSegment_00, IfcAlignmentSegment_01, IfcAlignmentSegment_02 } [minlen=5]

IfcAlignmentVertical -> IfcRelNests_2[dir=back]
IfcRelNests_2 -> { IfcAlignmentSegment_10, IfcAlignmentSegment_11, IfcAlignmentSegment_12 } [minlen=3]

IfcAlignmentCant -> IfcRelNests_3[dir=back]
IfcRelNests_3 -> { IfcAlignmentSegment_20, IfcAlignmentSegment_21, IfcAlignmentSegment_22 } [minlen=1]

IfcAlignmentSegment_00 -> IfcAlignmentHorizontalSegment_0
IfcAlignmentSegment_01 -> IfcAlignmentHorizontalSegment_1
IfcAlignmentSegment_02 -> IfcAlignmentHorizontalSegment_2

IfcAlignmentSegment_10 -> IfcAlignmentVerticalSegment_0
IfcAlignmentSegment_11 -> IfcAlignmentVerticalSegment_1
IfcAlignmentSegment_12 -> IfcAlignmentVerticalSegment_2

IfcAlignmentSegment_20 -> IfcAlignmentCantSegment_0
IfcAlignmentSegment_21 -> IfcAlignmentCantSegment_1
IfcAlignmentSegment_22 -> IfcAlignmentCantSegment_2



IfcAlignmentCantSegment_0       [label=IfcAlignmentCantSegment]
IfcAlignmentCantSegment_1       [label=IfcAlignmentCantSegment]
IfcAlignmentCantSegment_2       [label=IfcAlignmentCantSegment]
IfcAlignmentHorizontalSegment_0 [label=IfcAlignmentHorizontalSegment]
IfcAlignmentHorizontalSegment_1 [label=IfcAlignmentHorizontalSegment]
IfcAlignmentHorizontalSegment_2 [label=IfcAlignmentHorizontalSegment]
IfcAlignmentSegment_00          [label=IfcAlignmentSegment]
IfcAlignmentSegment_01          [label=IfcAlignmentSegment]
IfcAlignmentSegment_02          [label=IfcAlignmentSegment]
IfcAlignmentSegment_10          [label=IfcAlignmentSegment]
IfcAlignmentSegment_11          [label=IfcAlignmentSegment]
IfcAlignmentSegment_12          [label=IfcAlignmentSegment]
IfcAlignmentSegment_20          [label=IfcAlignmentSegment]
IfcAlignmentSegment_21          [label=IfcAlignmentSegment]
IfcAlignmentSegment_22          [label=IfcAlignmentSegment]
IfcAlignmentVerticalSegment_0   [label=IfcAlignmentVerticalSegment]
IfcAlignmentVerticalSegment_1   [label=IfcAlignmentVerticalSegment]
IfcAlignmentVerticalSegment_2   [label=IfcAlignmentVerticalSegment]
IfcRelNests_0                   [label=N, shape=circle]
IfcRelNests_1                   [label=N, shape=circle]
IfcRelNests_2                   [label=N, shape=circle]
IfcRelNests_3                   [label=N, shape=circle]


IfcAlignment -> IfcProductDefinitionShape           
IfcProductDefinitionShape -> IfcShapeRepresentation 
IfcShapeRepresentation -> IfcSegmentedReferenceCurve

IfcAlignment -> spacing1                 [style=invis]
spacing1 -> IfcProductDefinitionShape    [style=invis]
IfcProductDefinitionShape -> spacing2    [style=invis]
spacing2 -> IfcShapeRepresentation       [style=invis]
IfcShapeRepresentation -> spacing3       [style=invis]
spacing3 -> IfcSegmentedReferenceCurve   [style=invis]


{ IfcAlignment IfcProductDefinitionShape IfcShapeRepresentation IfcSegmentedReferenceCurve spacing1 spacing2 spacing3 rank=same }

IfcSegmentedReferenceCurve -> {  IfcCurveSegment_20, IfcCurveSegment_21, IfcCurveSegment_22  } [minlen=5]


IfcSegmentedReferenceCurve -> IfcGradientCurve

IfcGradientCurve -> {  IfcCurveSegment_10, IfcCurveSegment_11, IfcCurveSegment_12  }  [minlen=6]

IfcGradientCurve -> IfcCompositeCurve

IfcCompositeCurve -> {  IfcCurveSegment_00, IfcCurveSegment_01, IfcCurveSegment_02  } [minlen=7]

spacing1 [style=invis, width=2]
spacing2 [style=invis, width=2]
spacing3 [style=invis, width=2]


edge [constraint=false]

IfcAlignmentSegment_20 -> IfcCurveSegment_20 [style=dashed, label=Representation]
IfcAlignmentSegment_21 -> IfcCurveSegment_21 [style=dashed, label=Representation]
IfcAlignmentSegment_22 -> IfcCurveSegment_22 [style=dashed, label=Representation]
IfcAlignmentSegment_10 -> IfcCurveSegment_10 [style=dashed, label=Representation]
IfcAlignmentSegment_11 -> IfcCurveSegment_11 [style=dashed, label=Representation]
IfcAlignmentSegment_12 -> IfcCurveSegment_12 [style=dashed, label=Representation]
IfcAlignmentSegment_00 -> IfcCurveSegment_00 [style=dashed, label=Representation]
IfcAlignmentSegment_01 -> IfcCurveSegment_01 [style=dashed, label=Representation]
IfcAlignmentSegment_02 -> IfcCurveSegment_02 [style=dashed, label=Representation]

IfcCurveSegment_20 [label=IfcCurveSegment]
IfcCurveSegment_21 [label=IfcCurveSegment]
IfcCurveSegment_22 [label=IfcCurveSegment]
IfcCurveSegment_10 [label=IfcCurveSegment]
IfcCurveSegment_11 [label=IfcCurveSegment]
IfcCurveSegment_12 [label=IfcCurveSegment]
IfcCurveSegment_00 [label=IfcCurveSegment]
IfcCurveSegment_01 [label=IfcCurveSegment]
IfcCurveSegment_02 [label=IfcCurveSegment]

}
```

Reply:
> Where do you see this constraint encoded?

Perhaps I over stated my assumption. There are 4 CTs, the first 3 decompose the 3D space into two-2D representations that are combined to get a 3D result. I assumed the 4th CT, Segments, models the 3D space directly. The ParentCurve could be in a 2D horizontal xy plane, but this would be the Horizontal CT. Not sure if it make sense that ParentCurve be in a vertical sz plane without a horizontal component. 

I'm not sure what use case was imagined for the 4th CT but to me it looks like it would apply for as-built survey data which would be 3D points or segments

Reply:
Ok, I interpreted this differently. In my view the segment geometry is not really meant to be interesting in itself, but more to provide a more granular coupling between the business logic and the geometry. So the concept template just expresses that and doesn't really relate to a usecase. For surveying usecases it is possible to express an IfcAlignment representation simply as as 3d polyline.

(Or another possible top-level representation item for alignment: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcOffsetCurveByDistances.htm)